lib/xrp/XrpRpc.js

```diff
@@ -0,0 +1,237 @@
+const RippleAPI = require('ripple-lib').RippleAPI;
+const promptly = require('promptly');
+const util = require('util');
+const EventEmitter = require('events');
+
+const passwordPromptAsync = util.promisify(promptly.password);
+
+class XrpRpc {
+  constructor(config) {
+    this.config = config;
+    const {
+      port,
+      host,
+      protocol,
+      address,
+    } = config;
+    const connectionString = `${protocol}://${host}:${port}`;
+    this.rpc = new RippleAPI({
+      server: connectionString,
+    });
+    this.address = address;
+    this.emitter = new EventEmitter();
+  }
+
+  async asyncCall(method, args) {
+    return new Promise((resolve) => {
+      this.rpc.connect().then(() => {
+        return this.rpc[method](args).then((result) => {
+          this.rpc.disconnect().then(() => {
+            return resolve(result);
+          });
+        });
+      });
+    });
+  }
+
+  async asyncRequest(method, args) {
+    return new Promise((resolve) => {
+      this.rpc.connect().then(() => {
+        return this.rpc.request(method, args).then((result) => {
+          this.rpc.disconnect().then(() => {
+            return resolve(result);
+          });
+        });
+      });
+    });
+  }
+
+  async unlockAndSendToAddress({ address, amount, passphrase }) {
+    if (passphrase === undefined) {
+      passphrase = await passwordPromptAsync('> ');
+    }
+    console.warn('Unlocking for a single transaction.');
+    return await this.sendToAddress({ address, amount, passphrase });
+  }
+
+  async sendToAddress({ address, amount, passphrase, tag, invoiceID, secret }, alreadyConnected = false) {
+    return this.signTransaction({address, amount, passphrase, tag, invoiceID, secret}).then((signedTx) => {
+      return this.submitSignedTransaction({ signedTx }, alreadyConnected).then((txHash) => {
+        return txHash;
+      });
+    });
+  }
+
+  async getRawTransaction({ address, amount, tag, invoiceID, sourceAddress }) {
+    const payment = {
+      source: {
+        address: sourceAddress,
+        tag: tag || undefined,
+        maxAmount: {
+          value: amount.toString(),
+          currency: 'XRP'
+        }
+      },
+      destination: {
+        address: address,
+        tag: tag || undefined,
+        amount: {
+          value: amount.toString(),
+          currency: 'XRP'
+        }
+      },
+      invoiceID: invoiceID || undefined,
+    };
+    await this.rpc.connect();
+    let prepared = await this.rpc.preparePayment(sourceAddress, payment);
+    await this.rpc.disconnect();
+    return prepared;
+  }
+
+  async signTransaction({ address, amount, passphrase, tag, invoiceID, secret }) {
```

Comment:
We should add `alreadyConnected` to this method and connect if it isn't true. I get a `NotConnectedError` when doing `unlockAndSendToAddress`

Reply:
Fixed, added field and conditions for connect / disconnect